sw/otbn/crypto/field25519.s

```diff
@@ -0,0 +1,350 @@
+/* Copyright lowRISC contributors. */
+/* Licensed under the Apache License, Version 2.0, see LICENSE for details. */
+/* SPDX-License-Identifier: Apache-2.0 */
+
+/**
+ * This library contains arithmetic for the finite field modulo the prime
+ * p = 2^255-19, which is used for both the X25519 key exchange and the Ed25519
+ * signature scheme.
+ */
+
+
+/**
+ * Fully reduce a 510-bit number modulo p.
+ *
+ * Returns c = a mod p.
+ *
+ * Uses Solinas/pseudo-Mersenne reduction, which is based on the observation
+ * that if a large number x is split into two so that the lower 255 bits are x0
+ * and all bits above 255 are x1, then:
+ *
+ *   x = x0 + (2^255 * x1) \equiv (x0 + (19 * x1)) (mod p)
+ *
+ * If x is large, then x0 + 19 * x1 will be much smaller than x, because 19 <<
+ * 2^255. This step can be repeated as necessary until a conditional
+ * subtraction is enough to fully reduce.
+ *
+ * This routine runs in constant time.
+ *
+ * Flags: Flags have no meaning beyond the scope of this subroutine.
+ *
+ * @param[in]  w19: constant, w19 = 19
+ * @param[in]  [w21:w20]: a, number to be reduced (a < 2^510)
+ * @param[in]  MOD: p, modulus = 2^255 - 19
+ * @param[in]  w31: all-zero
```

Comment:
I assume there's a sensible architectural reason for this, but why doesn't this routine just zero w31 itself? Is this a code-size thing?

Is this a thing we do elsewhere (i.,e. would it be worth making part of the ABI globally)?

Reply:
It's pretty standard across all existing OTBN code to assume that w31 is always zero, and it indeed might be worth making it part of the ABI. Once I program the sign/verify/key-exchange entrypoints, I'll probably put that initialization there like we do for other OTBN code -- for code size reasons, like you've guessed.